include/ofi_mr.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2017 Intel Corporation, Inc. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include <inttypes.h>
+
+#include <fi.h>
+#include <fi_util.h>
+#include <fi_lock.h>
+#include <fi_list.h>
+
+#ifndef _OFI_MR_H_
+#define _OFI_MR_H_
+
+
+/*
+ * MR monitor
+ */
+
+struct ofi_mr_subscription;
+struct ofi_mr_monitor_eq;
+
+enum {
+	OFI_MR_MONITOR_EVENT_MAPPED	= 1 << 1,
+	OFI_MR_MONITOR_EVENT_UNMAPPED	= 1 << 2,
+};
+
+struct ofi_mr_monitor {
+	fastlock_t			lock;
+	struct dlist_entry		eq_list;
+	int				event_caps;
+
+	int (*subscribe)(struct ofi_mr_monitor *mr_monitor, void *addr,
+			 uint64_t len, struct ofi_mr_subscription *subscribe);
+	void (*unsubscribe)(struct ofi_mr_monitor *mr_monitor,
+			    struct ofi_mr_subscription *subscribe);
+	struct ofi_mr_subscription *(*get_event)(struct ofi_mr_monitor_eq *eq);
+};
+
+struct ofi_mr_monitor_eq {
+	struct ofi_mr_monitor		*mr_monitor;
+	struct dlist_entry		entry;
+	int				events;
+	fastlock_t			lock;
+	struct dlist_entry		list;
+};
+
+struct ofi_mr_subscription {
+	struct ofi_mr_monitor_eq	*eq;
+	int				events;
+	struct dlist_entry		entry;
+};
+
+int ofi_mr_monitor_init(struct ofi_mr_monitor *mr_monitor);
+void ofi_mr_monitor_finalize(struct ofi_mr_monitor *mr_monitor);
+void ofi_mr_monitor_eq_add(struct ofi_mr_monitor *mr_monitor,
+			   struct ofi_mr_monitor_eq *eq,
+			   int events);
```

Comment:
Do mr subscriptions inherit the events flag from the eq? If so, it would explain my question from below.

Reply:
I've designed it to work as follows:
User of MR monitor calls `ofi_mr_monitor_eq_add` and passes the events that user would expect from the `ofi_mr_monitor_get_event` call.

Answering on you question below:
> Also, is it assumed the cache will set the 'events' field?

the events flag should be set by ab underlying notifier when an event is generated.
i.e. if user request MAPPED and UNMAPPED event flags for the EQ, underlying notifier should set an appropriate event flag to a `ofi_mr_subscription` entry.

Reply:
> User of MR monitor calls ofi_mr_monitor_eq_add and passes the events that user would expect from the ofi_mr_monitor_get_event call.

That follows what I expected -- I want to confirm that it should be the responsibility of the underlying notifier to check the EQ event flags and discard generated events that the EQ hasn't requested. 

IE. if the underlying notifier generates MAPPED event, but the EQ only has the UNMAPPED enum value set, then the underlying implementation discards the event. 

If not, who is responsible for filtering the events?

Reply:
The underlying provider filters the events, but we can easily push this up to handle the case where the provider always gets all events